lib/ssl/src/tls_connection_1_3.erl

```diff
@@ -237,61 +237,94 @@ user_hello({call, From}, cancel, State) ->
+    {next_state, wait_sh, State#state{start_or_recv_from = From,
+                                      handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+     [{{timeout, handshake}, Timeout, close}]};
+user_hello({call, From}, {handshake_continue, NewOptions, Timeout},
+           #state{static_env = #static_env{role = server = Role},
+                  handshake_env = #handshake_env{handshake_continue = ClientVersions} = HSEnv,
+                  ssl_options = Options0} = State0) ->
+    Options= #{versions := Versions} = ssl:handle_options(NewOptions, Role, Options0),
+    State = ssl_gen_statem:ssl_config(Options, Role, State0),
+    case ssl_handshake:select_supported_version(ClientVersions, Versions) of
+        {3,4} ->
+            {next_state, start, State#state{start_or_recv_from = From,
+                                            handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+             [{{timeout, handshake}, Timeout, close}]};
+        undefined ->
+            ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), ?FUNCTION_NAME, State);
+        _Else ->
+            {next_state, hello, State#state{start_or_recv_from = From,
+                                            handshake_env = HSEnv#handshake_env{handshake_continue = continue}},
+             [{change_callback_module, tls_connection},
+              {{timeout, handshake}, Timeout, close}]}
+    end;
 user_hello(info, {'DOWN', _, _, _, _} = Event, State) ->
     ssl_gen_statem:handle_info(Event, ?FUNCTION_NAME, State);
 user_hello(_, _, _) ->
     {keep_state_and_data, [postpone]}.
 
 start(internal, #change_cipher_spec{}, State) ->
     tls_gen_connection:next_event(?FUNCTION_NAME, no_record, State);
-start(internal, #client_hello{extensions = Extensions} = Hello, 
-      #state{ssl_options = #{handshake := hello},
+start(internal, #client_hello{extensions = #{client_hello_versions :=
+                                                 #client_hello_versions{versions = ClientVersions}
+                                            }} = Hello,
+      #state{ssl_options = #{handshake := full}} = State) ->
+    case tls_record:is_acceptable_version({3,4}, ClientVersions) of
+        true ->
+            do_server_start(Hello, State);
+        false ->
+            ssl_gen_statem:handle_own_alert(?ALERT_REC(?FATAL, ?PROTOCOL_VERSION), ?FUNCTION_NAME, State)
+    end;
+start(internal, #client_hello{extensions = #{client_hello_versions :=
+                                                 #client_hello_versions{versions = ClientVersions}
+                                            }= Extensions},
+      #state{ssl_options = #{handshake := hello},  %% User requested pause to be able to react to hello extensions.
              start_or_recv_from = From,
-             handshake_env = HsEnv} = State) ->
+             handshake_env = #handshake_env{handshake_continue = Continue} = HSEnv} = State) when Continue =/= continue ->
```

Comment:
(low priority comment about variable naming style)

that part: `Continue =/= continue`

I understand it, it is probably fully functionally correct.
But from ease of code reading perspective ... maybe variable name could be adjusted ... `ContinuePhase`, `ContinueStage`, maybe some other word used in RFC ?

It might sound ridiculous, but I think that as long as it come at no cost and is doing the job - code should be readable for someone not knowing Erlang.

Reply:
It does not sound ridiculous :) I changed in an other way so that we get a positive test and do not need the variable!